keras_nlp/models/gpt2/gpt2_causal_lm.py

```diff
@@ -0,0 +1,333 @@
+        super().__init__(
+            inputs=inputs,
+            outputs=outputs,
+            include_preprocessing=preprocessor is not None,
+            **kwargs,
+        )
+
+        self._backbone = backbone
+        self._preprocessor = preprocessor
+
+    def preprocess_samples(self, x, y=None, sample_weight=None):
+        return self.preprocessor(x, y=y, sample_weight=sample_weight)
+
+    @property
+    def backbone(self):
+        """The associated `keras_nlp.models.GPT2Backbone`."""
+        return self._backbone
+
+    @property
+    def preprocessor(self):
+        """A `keras_nlp.models.GPT2CausalLMPreprocessor` for preprocessing."""
+        return self._preprocessor
+
+    @classproperty
+    def presets(cls):
+        return copy.deepcopy(backbone_presets)
+
+    @classmethod
+    def from_preset(
+        cls,
+        preset,
+        load_weights=True,
+        **kwargs,
+    ):
+        if "preprocessor" not in kwargs:
+            kwargs["preprocessor"] = GPT2CausalLMPreprocessor.from_preset(
+                preset
+            )
+
+        # Check if preset is backbone-only model.
+        if preset in GPT2Backbone.presets:
+            backbone = GPT2Backbone.from_preset(preset, load_weights)
+            return cls(backbone, **kwargs)
+
+        # Otherwise must be one of class presets.
+        # Currently no classifier-level presets, so we raise ValueError.
+        if preset not in cls.presets:
+            raise ValueError(
+                "`preset` must be one of "
+                f"""{", ".join(cls.presets)}. Received: {preset}."""
+            )
+
+    def _get_token_probability(self, prompt, mask):
+        model_inputs = {
+            "token_ids": prompt,
+            "padding_mask": mask,
+        }
+        return self(model_inputs)
+
+    def generate(
+        self,
+        prompt,
+        max_length,
+        end_token="<|endoftext|>",
+        sampler="top_k",
+    ):
+        """Generate text.
+
+        This method generates text based on given `prompt`. Generation will
+        continue until `max_length` is met, and all tokens generated after
+        `end_token` will be truncated.
+
+        Args:
+            prompt: a string, string Tensor or string RaggedTensor. The prompt
+                text for generation.
+            max_length: int. The max length of generated sequence.
+            end_token: string, defaults to "<|endoftext|>", which is the default
+                end token of GPT2. The token marking the end of the sequence,
+                tokens generated after the end token will be truncated.
+            sampler: a string or `keras_nlp.samplers.Sampler` instance. The
+                sampler to be used for text generation.
+        """
+        end_token_id = self.preprocessor.tokenizer.token_to_id(end_token)
+
+        if isinstance(sampler, str):
+            sampler = keras_nlp.samplers.get(sampler)
+        if hasattr(self, "jit_compile"):
```

Comment:
Why do we need this defensive hasattr? In the base model these are used directly, and these are documented properties on the model class...
https://www.tensorflow.org/api_docs/python/tf/keras/Model?version=nightly#attributes

Reply:
good catch, I wasn't aware that it is a property

Reply:
okay this is odd - for the `jit_compile` field, its default is not True/False, but None. Also we by default turn on `jit_compile` for samplers to make it faster, so I guess we need to keep the `jit_compile` attr check?

Reply:
I don't think we do actually! If we compile inside init in the future (as @jbischof is scoping current), we would just have `self.jit_compile == True` by default after the constructor is finished.

`hasattr` seems like the wrong thing here to deal with the none type, the attr always exists right, you will just get `None` when you read the value? If the goal is to remove `None` as an option, you might want `bool(self.jit_compile)` or something?

Reply:
I doubled checked - I think it's something we should fix on the Keras side, without calling `compile()`, `self.jit_compile` is an invalid call: ```AttributeError: 'GPT2CausalLM' object has no attribute 'jit_compile'```. 

We should let it able to return a default value before `compile()` is called.

Reply:
Oh interesting! `jit_compile` only became a property in late October, it probably hasn't made it's way to a stable release. So we would need to wait for our library to depend on 2.12 for that which is a ways off.

This looks good to me!